src/Build/Graph/ProjectGraph.cs

```diff
@@ -250,6 +252,84 @@ private void LoadGraph(string entryProjectFile, ProjectCollection projectCollect
             }
         }
 
+        // Using Tarjan's algorithm for detecting strongly connected components in a directed graph
+        // Reference: https://en.wikipedia.org/wiki/Strongly_connected_component
+        private int _index = 0;
+        private void DetectCyclesInGraph()
+        {
+            var indices = new Dictionary<ProjectGraphNode, int>();
+            var lowlinks = new Dictionary<ProjectGraphNode, int>();
+            var stack = new Stack<ProjectGraphNode>();
+            var existsOnStack = new HashSet<ProjectGraphNode>();
+            foreach (var projectNode in ProjectNodes)
+            {
+                if (!indices.ContainsKey(projectNode))
+                {
+                    if (IsStronglyConnected(projectNode, indices, lowlinks, stack, existsOnStack, out string scc))
+                    {
+                        throw new CircularDependencyException($"Exception occurred while creating depenendency graph: There is a circular dependency involving the following projects: {scc}");
+                    }
+                }
+            }
+        }
+
+        // indices stores the depth at which a node was found
+        // lowlink stores the smallest index of any node reachable from a node 
+        private bool IsStronglyConnected(
+            ProjectGraphNode projectNode,
+            Dictionary<ProjectGraphNode, int> indices,
+            Dictionary<ProjectGraphNode, int> lowlink,
+            Stack<ProjectGraphNode> stack,
+            HashSet<ProjectGraphNode> existsOnStack,
+            out string projectsInCycle
+            )
+        {
+            indices[projectNode] = _index;
+            lowlink[projectNode] = _index;
+            _index++;
+            stack.Push(projectNode);
+            existsOnStack.Add(projectNode);
+
+            foreach (var projectReference in projectNode.ProjectReferences)
+            {
+                if (!indices.ContainsKey(projectReference))
+                {
+                    if (IsStronglyConnected(projectReference, indices, lowlink, stack, existsOnStack, out projectsInCycle))
+                    {
+                        return true;
+                    }
+                    lowlink[projectNode] = Math.Min(lowlink[projectNode], lowlink[projectReference]);
+                }
+                else if (existsOnStack.Contains(projectReference))
+                {
+                    lowlink[projectNode] = Math.Min(lowlink[projectNode], indices[projectReference]);
+                }
+            }
+
+            if (lowlink[projectNode] == indices[projectNode])
```

Comment:
I believe you'll have to add a check for self-loops since this will fail on a self-referential dependency. Consider a graph such as
```
0 -> 1
⥁
```
where we've already visited ```1``` so the state looks like this:
```
indices: [0, 1]
lowLink: [0, 1]
stack: 0
existsOnStack: [T, F]
```

When iterating ```ProjectReferences``` from the first ```ProjectGraphNode```,  ```indices.ContainsKey(0) == true```and we won't add it to the stack again. Then although ```existsOnStack.Contains(0) == true``` and we would typically record a cycle, ```Math.min(lowlink[0], indices[0])``` will set ```lowlink[0]``` to ```0```. Then ```lowlink[projectNode] == indices[projectNode]``` is true, which triggers popping SCC from the stack. Since ```0``` is only on the stack once and ```0 == 0```, ```countOfProjectsInCycle == 1``` although a cycle exists.